dragonphy/fpga_models/analog_slice.py

```diff
@@ -108,25 +109,63 @@ def __init__(self, filename=None, **system_values):
                 rst=m.rst
             )
 
-        # Compute the evaluation time for this slice
-        t_samp_pre = m.bind_name(
-            't_samp_pre',
-            ((m.slice_offset*(1<<system_values['pi_ctl_width'])) + m.pi_ctl_sample)
-            / ((2.0**system_values['pi_ctl_width'])*system_values['freq_rx'])
+        # Compute the delay due to the PI control code
+        delay_amt_pre = m.bind_name(
+            'delay_amt_pre',
+            m.pi_ctl_sample / ((2.0**system_values['pi_ctl_width'])*system_values['freq_rx'])
         )
 
         # Add jitter to the sampling time
         if system_values['use_jitter']:
-            t_samp_jitter = m.set_gaussian_noise(
-                't_samp_jitter',
+            # create a signal to represent jitter
+            delay_amt_jitter = m.set_gaussian_noise(
+                'delay_amt_jitter',
                 std=m.jitter_rms,
                 lfsr_init=m.jitter_seed,
                 clk=m.clk,
+                ce=m.sample_ctl,
                 rst=m.rst
             )
-            t_samp = m.bind_name('t_samp', t_samp_pre + t_samp_jitter)
+
+            # add jitter to the delay amount (which might possibly yield a negative value)
+            delay_amt_noisy = m.bind_name('delay_amt_noisy', delay_amt_pre + delay_amt_jitter)
+
+            # make the delay amount non-negative
+            delay_amt = m.bind_name('delay_amt', if_(delay_amt_noisy >= 0.0, delay_amt_noisy, 0.0))
         else:
-            t_samp = t_samp_pre
+            delay_amt = delay_amt_pre
```

Comment:
what follows is a bunch of code to handle scenarios where the sampling point occurs after the clk_adc edge.  for example, when the last sample was "early" and this sample is "late", we have to keep the output data constant and pass the sampling time onto the next clk_adc cycle.  this actually makes a significant difference at low BER